pkg/datapath/linux/sysctl/sysctl.go

```diff
@@ -0,0 +1,208 @@
+func (sysctl *Sysctl) ApplySettings(sysSettings []tables.Sysctl) error {
+	txn := sysctl.db.WriteTxn(sysctl.settings)
+	for _, s := range sysSettings {
+		_, _, _ = sysctl.settings.Insert(txn, s.WithStatus(reconciler.StatusPending()))
+	}
+	txn.Commit()
+
+	var errs []error
+	for _, s := range sysSettings {
+		errs = append(errs, sysctl.waitForReconciliation(s.Name))
+	}
+
+	return errors.Join(errs...)
+}
+
+// Read reads the given sysctl parameter.
+func (sysctl *Sysctl) Read(name string) (string, error) {
+	val, err := readSysctl(sysctl.fs, sysctl.procFs, name)
+	if err != nil {
+		return "", err
+	}
+
+	return val, nil
+}
+
+// ReadInt reads the given sysctl parameter, return an int64 value.
+func (sysctl *Sysctl) ReadInt(name string) (int64, error) {
+	val, err := sysctl.Read(name)
+	if err != nil {
+		return -1, err
+	}
+
+	v, err := strconv.ParseInt(val, 10, 64)
+	if err != nil {
+		return -1, err
+	}
+
+	return v, nil
+}
+
+// parameterElemRx matches an element of a sysctl parameter.
+var parameterElemRx = regexp.MustCompile(`\A[-0-9_a-z]+\z`)
+
+// parameterPath returns the path to the sysctl file for parameter name.
+func parameterPath(procFs, name string) (string, error) {
+	elems := strings.Split(name, ".")
+	for _, elem := range elems {
+		if !parameterElemRx.MatchString(elem) {
+			return "", fmt.Errorf("invalid sysctl parameter: %q", name)
+		}
+	}
+	return filepath.Join(append([]string{procFs, "sys"}, elems...)...), nil
+}
+
+func writeSysctl(fs afero.Fs, procFs, name, value string) error {
+	path, err := parameterPath(procFs, name)
+	if err != nil {
+		return err
+	}
+
+	f, err := fs.OpenFile(path, os.O_RDWR, 0644)
+	if err != nil {
+		return fmt.Errorf("could not open the sysctl file %s: %s", path, err)
+	}
+	defer f.Close()
+
+	if _, err := io.WriteString(f, value); err != nil {
+		return fmt.Errorf("could not write to the systctl file %s: %s",
+			path, err)
+	}
+	return nil
+}
+
+func readSysctl(fs afero.Fs, procFs, name string) (string, error) {
+	path, err := parameterPath(procFs, name)
+	if err != nil {
+		return "", err
+	}
+
+	f, err := fs.Open(path)
+	if err != nil {
+		return "", fmt.Errorf("could not open the sysctl file %s: %s", path, err)
+	}
+	defer f.Close()
+
+	val, err := safeio.ReadAllLimit(f, safeio.KB)
+	if err != nil {
+		return "", fmt.Errorf("could not read the systctl file %s: %s", path, err)
+	}
+
+	return strings.TrimRight(string(val), "\n"), nil
+}
+
+func (sysctl *Sysctl) waitForReconciliation(name string) error {
+	obj, _, watch, _ := sysctl.settings.FirstWatch(sysctl.db.ReadTxn(), tables.SysctlNameIndex.Query(name))
+	if obj.Status.Kind == reconciler.StatusKindDone {
+		// already reconciled
+		return nil
+	}
+
+	t := time.NewTimer(time.Second)
+	defer t.Stop()
+	select {
+	case <-t.C:
+		return fmt.Errorf("timeout waiting for parameter %s reconciliation", name)
+	case <-watch:
+	}
+	if obj.Status.Kind == reconciler.StatusKindError {
+		return errors.New(obj.Status.Error)
```

Comment:
Hmm this makes me wonder if `Status.Error` should be `error`. I think the reason why I went with `string` was to make the default JSON dumping look reasonable, but perhaps it'd be nicer to keep it as `error` and just have custom JSON marshalling for `Status`...

Reply:
I think having an `error` + custom JSON marshalling would be better. I expect many uses of `Status.Error` as the reconciler is used more.